packages/clay-core/src/tree-view/TreeView.tsx

```diff
@@ -0,0 +1,122 @@
+/**
+ * SPDX-FileCopyrightText: © 2021 Liferay, Inc. <https://liferay.com>
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import classNames from 'classnames';
+import React from 'react';
+import {DndProvider} from 'react-dnd';
+import HTML5Backend from 'react-dnd-html5-backend';
+
+import {ChildrenFunction, Collection, ICollectionProps} from './Collection';
+import {TreeViewGroup} from './TreeViewGroup';
+import {TreeViewItem, TreeViewItemStack} from './TreeViewItem';
+import {Icons, TreeViewContext} from './context';
+import {ITreeProps, useTree} from './useTree';
+
+function generateItemIndices<T>(
+	items: Array<T> | Record<string, T> | undefined
+) {
+	if (!items) {
```

Comment:
@matuzalemsteles  About this, I know why we defined items as `Array<T> | Record<string, T> | undefined` (we talked about in either #4218 or #4222), but there's got to be a better way to deal with these union types, I need to clean up.